community/kernel/src/main/java/org/neo4j/kernel/impl/store/NeoStore.java

```diff
@@ -900,12 +900,32 @@ public long nextCommittingTransactionId()
     @Override
     public void transactionCommitted( long transactionId, long checksum )
     {
-        if ( lastCommittedTx.offer( transactionId, checksum ) )
+        checkInitialized( lastCommittingTxField.get() );
+        HighestCommittedTransaction current = highestCommittedTransaction.get();
+        if ( current.transactionId() < transactionId )
         {
-            long[] transactionData = lastCommittedTx.get();
-            setRecord( Position.LAST_TRANSACTION_ID, transactionData[0] );
-            setRecord( Position.LAST_TRANSACTION_CHECKSUM, transactionData[1] );
-            lastTransactionChecksum = checksum;
+            HighestCommittedTransaction update = new HighestCommittedTransaction( transactionId, checksum );
+            while ( !highestCommittedTransaction.compareAndSet( current, update ) )
+            {
+                current = highestCommittedTransaction.get();
+                // let's check if our update is still needed
+                if ( current.transactionId() >= transactionId )
+                {
+                    return;
+                }
+            }
+
+            /*
+             * We need to synchronize here in order to guarantee that the two field are written consistently
+             * together. Note that having the exclusive lock on tha page is not enough for 2 reasons:
+             * 1. the records might be in different pages
```

Comment:
Actually, for the neostore all the records will be on the same page, and even if this store would later grow to span multiple pages we can make sure these two records are next to each other so that they land on the same page anyway